# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: accessible/base/FocusManager.cpp
# Commit: 2010ac6007c8
# Full Hash: 2010ac6007c8ef7684a5fbef2903fc42aa3e241d
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2021-03-25 03:42:26
# Description:
#   Bug 1691930 - Check that FocusedChild is bound to a doc in RecvFocusedChild. r=Jamie
#   
#   Added an assert in focus manager. Hopefully fuzzers will help us find
#   cases where the active item is defunct, if that is indeed what is
#   happening.
# ==============================================================================

diff -r 882197c37c04 -r 2010ac6007c8 accessible/base/FocusManager.cpp
--- a/accessible/base/FocusManager.cpp	Wed Mar 24 18:10:34 2021 +0200
+++ b/accessible/base/FocusManager.cpp	Wed Mar 24 16:24:47 2021 +0000
@@ -28,7 +28,14 @@
 FocusManager::~FocusManager() {}
 
 LocalAccessible* FocusManager::FocusedAccessible() const {
-  if (mActiveItem) return mActiveItem;
+  if (mActiveItem) {
+    if (mActiveItem->IsDefunct()) {
+      MOZ_ASSERT_UNREACHABLE("Stored active item is unbound from document");
+      return nullptr;
+    }
+
+    return mActiveItem;
+  }
 
   nsINode* focusedNode = FocusedDOMNode();
   if (focusedNode) {